browser-polyfill.js

```diff
@@ -0,0 +1,340 @@
+      return function asyncFunctionWrapper(target, ...args) {
+        if (args.length < metadata.minArgs) {
+          throw new Error(`Expected at least ${metadata.minArgs} arguments for ${name}(), got ${args.length}`);
+        }
+
+        if (args.length > metadata.maxArgs) {
+          throw new Error(`Expected at most ${metadata.maxArgs} arguments for ${name}(), got ${args.length}`);
+        }
+
+        return new Promise((resolve, reject) => {
+          target[name](...args, makeCallback({resolve, reject}));
+        });
+      };
+    };
+
+    /**
+     * Wraps an existing method of the target object, so that calls to it are
+     * intercepted by the given wrapper function. The wrapper function receives,
+     * as its first argument, the original `target` object, followed by each of
+     * the arguments passed to the orginal method.
+     *
+     * @param {object} target
+     *        The original target object that the wrapped method belongs to.
+     * @param {function} method
+     *        The method being wrapped. This is used as the target of the Proxy
+     *        object which is created to wrap the method.
+     * @param {function} wrapper
+     *        The wrapper function which is called in place of a direct invocation
+     *        of the wrapped method.
+     *
+     * @returns {Proxy<function>}
+     *        A Proxy object for the given method, which invokes the given wrapper
+     *        method in its place.
+     */
+    const wrapMethod = (target, method, wrapper) => {
+      return new Proxy(method, {
+        apply(targetMethod, thisObj, args) {
+          return wrapper.call(thisObj, target, ...args);
+        },
+      });
+    };
+
+    let hasOwnProperty = Function.call.bind(Object.prototype.hasOwnProperty);
+
+    /**
+     * Wraps an object in a Proxy which intercepts and wraps certain methods
+     * based on the given `wrappers` and `metadata` objects.
+     *
+     * @param {object} target
+     *        The target object to wrap.
+     *
+     * @param {object} [wrappers = {}]
+     *        An object tree containing wrapper functions for special cases. Any
+     *        function present in this object tree is called in place of the
+     *        method in the same location in the `target` object tree. These
+     *        wrapper methods are invoked as described in {@see wrapMethod}.
+     *
+     * @param {object} [metadata = {}]
+     *        An object tree containing metadata used to automatically generate
+     *        Promise-based wrapper functions for asynchronous. Any function in
+     *        the `target` object tree which has a corresponding metadata object
+     *        in the same location in the `metadata` tree is replaced with an
+     *        automatically-generated wrapper function, as described in
+     *        {@see wrapAsyncFunction}
+     *
+     * @returns {Proxy<object>}
+     */
+    const wrapObject = (target, wrappers = {}, metadata = {}) => {
```

Comment:
nit

There are a lot of things called `wrapper` and `wrappers` here, how about renaming this `wrappers` parameter into something like "specialAPIWrappers" and the `wrapper` defined in the body into something like `asyncFunctionWrapper`?
(it would make it faster to visually recognize where a wrapper represents a special API or an async function)